### PR TITLE
doc(gorgone) add documentation for developper and pullwss connexion mode

### PR DESCRIPTION
**Pull request by Evan-Adam**
## Description
Add documentation for gorgone.

**Fixes** # MON-37896

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [X] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 22.04.x
- [ ] 22.10.x
- [ ] 23.04.x
- [ ] 23.10.x
- [X] 24.04.x (master)

<h2> How this pull request can be tested ? </h2>

Please describe the **procedure** to verify that the goal of the PR is matched. Provide clear instructions so that it can be **correctly tested**.

Any **relevant details** of the configuration to perform the test should be added.

#### Community contributors & Centreon team

- [ ] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: centreon-gorgone/docs/poller_pullwss_configuration.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
+# Architecture
+
+We are showing how to configure gorgone to manage that architecture:
+
+```text
+Central server <------- Distant Poller
+```
+contrary to pull module, the communication is entierely done on http(s) websocket.
```
[the remark this answers is not in the document]
Comment (Evan-Adam) [reply centreon-gorgone/docs/poller_pullwss_configuration.md:8]: not done on all the rest of the gorgone doc.

File: centreon-gorgone/docs/modules/core/proxy.md [hunk 2]
```diff
@@ -13,13 +13,28 @@ A SSH client library make routing to non-gorgoned nodes possible.
...
+| ssl           | should connection be made over tls/ssl or not                                                  | `false`       |
+| ssl_cert_file | Path to a ssl certificate file. required if ssl: true                                          |               |
+| ssl_key_file  | Path to a ssl key file associated to the certificate already configured. required if ssl: true |               |
+| passphrase    | No idea, maybe an optionnal pasphrase for the ssl key.                                         |               |
+| token         | Allow to authenticate node. Is required to enable http server.                                 |               |
+| address       | Address to listen to. can be 0.0.0.0 to listen on all ipv4 address                             |               |
+| port          | tcp port to listen to.                                                                         |               |
+
+
 #### Example
```
[the remark this answers is not in the document]
Comment (Evan-Adam) [reply centreon-gorgone/docs/modules/core/proxy.md:37]: this is done in all gorgone doc.

File: centreon-gorgone/docs/modules.md [hunk 3]
```diff
@@ -19,3 +19,87 @@ List of the available modules:
...
+
+### event()
+
+Reads data from ZMQ socket, and acts on it, generally by launching an action_* method to process the event.
+
+module.pm parent class has an event() method, so it's not mandatory to implement it. 
+
+### action_*()
+
+Method called by event() when a ZMQ message is found.\
```
Comment (omercier) [inline centreon-gorgone/docs/modules.md:103]:
```suggestion
Method called by event() when a ZMQ message is found.

```

File: centreon-gorgone/docs/modules/centreon/autodiscovery.md [hunk 4]
```diff
@@ -316,3 +316,26 @@ curl --request POST "https://hostname:8443/api/centreon/autodiscovery/services"
...
+
+Dotted lines mean a ZMQ message is sent. Direct lines mean the function is called normally.
+
+Each column represents a Linux thread, as Gorgone is multi process.
+
+For each ZMQ message, names are described in the [events section](#events) of each module,
+and for putlog the second part is the 'code' used by gorgone-autodiscovery
+and defined as constant in the [class.pm](../../../gorgone/modules/centreon/autodiscovery/class.pm) file.
+
+The gorgone-action module does not send the result directly to the calling module. It sends a putlog message instead, processed by core.\
```
Comment (omercier) [inline centreon-gorgone/docs/modules/centreon/autodiscovery.md:336]:
```suggestion
The gorgone-action module does not send the result directly to the calling module. It sends a putlog message instead, processed by core.

```

File: centreon-gorgone/docs/modules/centreon/autodiscovery.md [hunk 5]
```diff
@@ -316,3 +316,26 @@ curl --request POST "https://hostname:8443/api/centreon/autodiscovery/services"
...
 }"
 ```
+
+### Developer manual
+
+This module heavily uses the gorgone-action module to work.
+
+Here is a diagram of how these modules interact:
+
+![image](./centreon-gorgone-autodiscovery-archi.png)
```
Comment (omercier) [inline centreon-gorgone/docs/modules/centreon/autodiscovery.md:326]:
```suggestion
![image](./centreon-gorgone-autodiscovery-archi.jpg)

```

File: centreon-gorgone/docs/modules/centreon/autodiscovery.md [hunk 6]
```diff
@@ -316,3 +316,26 @@ curl --request POST "https://hostname:8443/api/centreon/autodiscovery/services"
...
+Each column represents a Linux thread, as Gorgone is multi process.
+
+For each ZMQ message, names are described in the [events section](#events) of each module,
+and for putlog the second part is the 'code' used by gorgone-autodiscovery
+and defined as constant in the [class.pm](../../../gorgone/modules/centreon/autodiscovery/class.pm) file.
+
+The gorgone-action module does not send the result directly to the calling module. It sends a putlog message instead, processed by core.\
+Core keeps track of every module waiting for a particular event (use library.pm::addlistener to show interest in an event)\
+
+and dispatch another message to the waiting module.
```
Comment (omercier) [inline centreon-gorgone/docs/modules/centreon/autodiscovery.md:339]:
```suggestion
Core keeps track of every module waiting for a particular event (use library.pm::addlistener to show interest in an event)
and dispatch another message to the waiting module.

```

File: centreon-gorgone/docs/modules/core/proxy.md [hunk 7]
```diff
@@ -13,13 +13,27 @@ A SSH client library make routing to non-gorgoned nodes possible.
...
 | Directive            | Description                                                         | Default value |
-| :------------------- | :------------------------------------------------------------------ | :------------ |
+|:---------------------|:--------------------------------------------------------------------|:--------------|
 | pool                 | Number of childs to instantiate to process events                   | `5`           |
 | synchistory_time     | Time in seconds between two logs synchronisation                    | `60`          |
 | synchistory_timeout  | Time in seconds before logs synchronisation is considered timed out | `30`          |
 | ping                 | Time in seconds between two node pings                              | `60`          |
 | pong_discard_timeout | Time in seconds before a node is considered dead                    | `300`         |
 
+this part of the configuration is only used if some poller must connect with pullwss module.
```
Comment (omercier) [inline centreon-gorgone/docs/modules/core/proxy.md:23]:
```suggestion
This part of the configuration is only used if some poller must connect with pullwss module.

```

File: centreon-gorgone/docs/modules/core/proxy.md [hunk 8]
```diff
@@ -13,13 +13,27 @@ A SSH client library make routing to non-gorgoned nodes possible.
...
+| httpserver    | array containing all below configuration for pullwss connnection                               | no value.     |
+| enable        | boolean if http server should be enabled                                                       | `false`       |
+| ssl           | should connection be made over tls/ssl or not                                                  | `false`       |
+| ssl_cert_file | Path to a ssl certificate file. required if ssl: true                                          |               |
+| ssl_key_file  | Path to a ssl key file associated to the certificate already configured. required if ssl: true |               |
+| passphrase    | No idea, maybe an optionnal pasphrase for the ssl key.                                         |               |
+| token         | Allow to authenticate node. It is required to enable the http server.                          |               |
+| address       | Address to listen to. can be 0.0.0.0 to listen on all IPv4 address                             |               |
+| port          | tcp port to listen to.                                                                         |               |
```
Comment (omercier) [inline centreon-gorgone/docs/modules/core/proxy.md:35]:
```suggestion
| httpserver    | Array containing all below configuration for pullwss connnection                               | no value.     |
| enable        | Boolean if http server should be enabled                                                       | `false`       |
| ssl           | Should connection be made over tls/ssl or not                                                  | `false`       |
| ssl_cert_file | Path to a ssl certificate file. required if ssl: true                                          |               |
| ssl_key_file  | Path to a ssl key file associated to the certificate already configured. required if ssl: true |               |
| passphrase    | No idea, maybe an optionnal pasphrase for the ssl key.                                         |               |
| token         | Allow to authenticate node. It is required to enable the http server.                          |               |
| address       | Address to listen to. can be 0.0.0.0 to listen on all IPv4 address                             |               |
| port          | TCP port to listen to.                                                                         |               |

```

File: centreon-gorgone/docs/modules/core/proxy.md [hunk 9]
```diff
@@ -31,6 +45,13 @@ synchistory_time: 60
 synchistory_timeout: 30
 ping: 60
 pong_discard_timeout: 300
+httpserver:  # this is used only if you want to configure pullwss nodes. to work you have to add register module and configure a config file.
```
Comment (omercier) [inline centreon-gorgone/docs/modules/core/proxy.md:48]:
```suggestion
httpserver:  # This is used only if you want to configure pullwss nodes. To make it work you have to add a *register module* file and set a configuration file.

```

**Review by omercier — CHANGES_REQUESTED** [review]

**Review by omercier — DISMISSED** [review]

**Review by omercier — DISMISSED** [review]

**Review by lpinsivy — CHANGES_REQUESTED** [review]

**Review by lpinsivy — DISMISSED** [review]

File: centreon-gorgone/docs/modules.md [hunk 10]
```diff
@@ -19,3 +19,81 @@ List of the available modules:
...
+### broadcast()
+
+### create_child()
+
+Not strictly required, but present every time, used to instantiate a new child process by the init() function.\
+Inside the child process, a class.pm object is created and the class->run method is started.
+
+## class.pm
+
+This class must inherit module.pm package.
```
Comment (cg-tw) [inline centreon-gorgone/docs/modules.md:65]:
```suggestion
This class must inherit the module.pm package.

```

File: centreon-gorgone/docs/modules.md [hunk 11]
```diff
@@ -19,3 +19,81 @@ List of the available modules:
...
+
+This object is most of the time (maybe all ?) a singleton.
```
Comment (cg-tw) [inline centreon-gorgone/docs/modules.md:67]:
```suggestion
This object is most of the time a singleton (maybe every time).

```

File: centreon-gorgone/docs/modules.md [hunk 12]
```diff
@@ -19,3 +19,81 @@ List of the available modules:
...
+
+It will be created by hook.pm when starting the module.
+This is the workhorse that will process all events.
+
+It seems like none of these methods will be called by gorgone-core, so naming is not required to follow this convention.
+
+(please keep the code base consistent if you make a new module).
```
Comment (cg-tw) [inline centreon-gorgone/docs/modules.md:74]:
```suggestion
(Please keep the code base consistent if you make a new module).

```

File: centreon-gorgone/docs/modules/core/pullwss.md [hunk 13]
```diff
@@ -0,0 +1,41 @@
+# Pullwss
+
+## Description
+
+This module should be used on remote nodes where the connection has to be http/https and must be opened from the node to the Central Gorgone.
+
+This module require proxy and register module to be configured on the central Gorgone.
+Register will allow to keep the state of every poller, and find out the connexion mode.
```
[the remark this answers is not in the document]
Comment (Evan-Adam) [reply centreon-gorgone/docs/modules/core/pullwss.md:8]: Register module allow gorgone to keep the state, as this documentation is for futur developer, I think we don't need the 'you'

File: centreon-gorgone/docs/modules/core/pullwss.md [hunk 14]
```diff
@@ -0,0 +1,41 @@
+# Pullwss
+
+## Description
+
+This module should be used on remote nodes where the connection has to be http/https and must be opened from the node to the Central Gorgone.
+
+This module require proxy and register module to be configured on the central Gorgone.
+Register will allow to keep the state of every poller, and find out the connexion mode. 
+Proxy have to bind to a tcp port for pullwss module to connect to.
```
[the remark this answers is not in the document]
Comment (Evan-Adam) [reply centreon-gorgone/docs/modules/core/pullwss.md:9]:
```suggestion
Proxy module has to bind to a tcp port for pullwss module to connect to.
```

File: centreon-gorgone/docs/modules/core/pullwss.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (Evan-Adam) [reply centreon-gorgone/docs/modules/core/pullwss.md:8]:
```suggestion
Register Module will allow Gorgone to keep the state of every poller, and find out the connexion mode. 
```

File: centreon-gorgone/docs/modules/core/pullwss.md [hunk 15]
```diff
@@ -0,0 +1,41 @@
+# Pullwss
+
+## Description
+
+This module should be used on remote nodes where the connection has to be HTTP/HTTPS and must be opened from the node to the Central Gorgone.
+
+This module requires proxy and register module to be configured on the central Gorgone.
+Register Module will allow Gorgone to keep the state of every poller, and find out the connexion mode. 
+Proxy module has to bind to a tcp port for the pullwss module to connect to.
```
Comment (cg-tw) [inline centreon-gorgone/docs/modules/core/pullwss.md:9]:
```suggestion
The proxy module has to bind to a tcp port for the pullwss module to connect to.
```

File: centreon-gorgone/docs/modules/core/pullwss.md [hunk 16]
```diff
@@ -0,0 +1,41 @@
+# Pullwss
+
+## Description
+
+This module should be used on remote nodes where the connection has to be HTTP/HTTPS and must be opened from the node to the Central Gorgone.
+
+This module requires proxy and register module to be configured on the central Gorgone.
+Register Module will allow Gorgone to keep the state of every poller, and find out the connexion mode.
```
Comment (cg-tw) [inline centreon-gorgone/docs/modules/core/pullwss.md:8]:
```suggestion
The register Module will allow Gorgone to keep the state of every poller, and find out the connexion mode. 
```

**Comment by Evan-Adam** [comment]
@cg-tw  can you check this ?

**Review by cg-tw — APPROVED** [review]

**Review by lucie-dubrunfaut — APPROVED** [review]

**Comment by Evan-Adam** [comment]
@coderabbitai pause

**Pull request merged**